packages/app/src/systems/Chains/eth/machines/assetsMachine.tsx

```diff
@@ -0,0 +1,180 @@
+    tsTypes: {} as import('./assetsMachine.typegen').Typegen0,
+    schema: {
+      context: {} as MachineContext,
+      services: {} as MachineServices,
+      events: {} as AssetListMachineEvents,
+    },
+    predictableActionArguments: true,
+    id: '(machine)',
+    initial: 'setListedAssets',
+    states: {
+      setListedAssets: {
+        tags: ['loading'],
+        invoke: {
+          src: 'setListedAssets',
+          onDone: [
+            {
+              target: 'fetchingAssets',
+            },
+          ],
+        },
+      },
+      fetchingAssets: {
+        tags: ['loading'],
+        invoke: {
+          src: 'fetchAssets',
+          onDone: [
+            {
+              actions: ['assignAssets'],
+              target: 'idle',
+            },
+          ],
+        },
+      },
+      idle: {
+        on: {
+          ADD_ASSET: {
+            target: 'adding',
+          },
+          REMOVE_ASSET: {
+            target: 'removing',
+          },
+        },
+      },
+      adding: {
+        tags: ['loading'],
+        invoke: {
+          src: 'addAsset',
+          data: {
+            input: (_: MachineContext, ev: AssetListMachineEvents) => ev.input,
+          },
+          onDone: {
+            target: 'fetchingAssets',
+          },
+        },
+      },
+      removing: {
+        tags: ['loading'],
+        invoke: {
+          src: 'removeAsset',
+          data: {
+            input: (_: MachineContext, ev: AssetListMachineEvents) => ev.input,
+          },
+          onDone: [
+            {
+              target: 'fetchingAssets',
+            },
+          ],
+        },
+      },
+    },
+  },
+  {
+    actions: {
+      assignAssets: assign({
+        assetList: (_, ev) => ev.data,
+      }),
+    },
+    services: {
+      setListedAssets: FetchMachine.create<null, void>({
+        showError: true,
+        async fetch() {
+          await Promise.all(
+            AssetList.map((asset) =>
+              AssetService.upsertAsset({
+                data: {
+                  ...asset,
+                  image: asset.image,
+                },
+              })
+            )
+          );
+        },
+      }),
+      fetchAssets: FetchMachine.create<null, BridgeAsset[]>({
+        showError: true,
+        async fetch() {
+          return AssetService.getAssets();
+        },
+      }),
+      addAsset: FetchMachine.create<{ asset: BridgeAsset }, boolean>({
+        showError: true,
+        async fetch({ input }) {
+          if (!input) {
+            throw new Error('No input to add asset');
+          }
+
+          await AssetService.addAsset({ data: input.asset });
+          return true;
+        },
+      }),
+      removeAsset: FetchMachine.create<
+        {
+          address: string;
+        },
+        boolean
+      >({
+        showError: true,
+        async fetch({ input }) {
+          if (!input?.address) {
+            throw new Error('Missing data');
+          }
+
```

Comment:
can we move this validation to inside the `AssetService.removeAsset` method?

Reply:
is this better?
```
if (!input) {
  throw new Error('Missing data');
}
```

Reply:
we can have both ? I think validating the address is a good idea, but should be done inside the service as it's a must have for the `removeASset` action to work